File: crates/a2ui-renderer/src/path_resolver.rs
use crate::error::RenderResult;
use a2ui_core::prelude::*;
use serde_json::Value;

/// 路径解析引擎
/// 支持绝对路径、相对路径（集合作用域）和 DynamicValue 解析
#[derive(Debug, Clone)]
pub struct PathResolver {
    data_model: DataModel,
    /// 集合作用域栈（用于解析相对路径）
    scope_stack: Vec<Scope>,
}

/// 作用域类型
#[derive(Debug, Clone)]
pub enum Scope {
    /// 根作用域
    Root,
    /// 集合作用域（数组迭代中的当前项）
    Collection { base_path: String, index: usize },
}

impl PathResolver {
    /// 从 DataModel 创建
    pub fn new(data_model: DataModel) -> Self {
        Self {
            data_model,
            scope_stack: vec![Scope::Root],
        }
    }

    /// 解析路径（支持绝对和相对路径）
    pub fn resolve(&self, path: &str) -> Option<Value> {
        if path.starts_with('/') {
            // 绝对路径
            self.data_model.get(path).cloned()
        } else if path == "@index" {
            // 系统保留变量
            self.current_index().map(|i| json!(i))
        } else {
            // 相对路径：在当前作用域下解析
            self.resolve_relative(path)
        }
    }

    /// 解析相对路径
    pub fn resolve_relative(&self, relative: &str) -> Option<Value> {
        let base = self.current_base_path();
        if base.is_empty() {
            self.data_model.get(&format!("/{}", relative)).cloned()
        } else {
            let idx = self.current_index().unwrap_or(0);
            let full = format!("{}/{}/{}", base, idx, relative);
            self.data_model.get(&full).cloned()
        }
    }

    /// 将相对路径转换为绝对路径
    pub fn resolve_to_absolute(&self, base: &str, relative: &str, index: usize) -> String {
        if relative.starts_with('/') {
            relative.to_string()
        } else {
            format!("{}/{}/{}", base, index, relative)
        }
    }

    /// 解析 DynamicValue
    pub fn resolve_dynamic<T>(&self, dynamic: &DynamicValue<T>) -> RenderResult<Value>
    where
        T: Clone + Into<Value>,
    {
        match dynamic {
            DynamicValue::Literal(v) => Ok(v.clone().into()),
            DynamicValue::Path { path } => self.resolve(path).ok_or_else(|| {
                crate::error::RendererError::SurfaceNotFound(format!("path not found: {}", path))
            }),
            DynamicValue::FunctionCall { call, .. } => Err(
                crate::error::RendererError::FunctionNotAvailable(call.clone()),
            ),
        }
    }

    /// 获取当前作用域的 JSON Pointer 基础路径
    fn current_base_path(&self) -> String {
        if let Some(scope) = self.scope_stack.iter().next_back() {
            match scope {
                Scope::Collection { base_path, .. } => return base_path.clone(),
                Scope::Root => {}
            }
        }
        String::new()
    }

    /// 获取当前索引
    fn current_index(&self) -> Option<usize> {
        if let Some(scope) = self.scope_stack.iter().next_back() {
            match scope {
                Scope::Collection { index, .. } => return Some(*index),
                Scope::Root => {}
            }
        }
        None
    }

    /// 进入集合作用域
    pub fn enter_collection(&mut self, base_path: impl Into<String>, index: usize) {
        self.scope_stack.push(Scope::Collection {
            base_path: base_path.into(),
            index,
        });
    }

    /// 退出集合作用域
    pub fn exit_collection(&mut self) {
        if self.scope_stack.len() > 1 {
            self.scope_stack.pop();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_resolve_absolute_path() {
        let dm = DataModel::new(json!({"user": {"name": "Alice"}}));
        let resolver = PathResolver::new(dm);
        assert_eq!(resolver.resolve("/user/name"), Some(json!("Alice")));
    }

    #[test]
    fn test_resolve_nonexistent() {
        let dm = DataModel::new(json!({}));
        let resolver = PathResolver::new(dm);
        assert!(resolver.resolve("/missing").is_none());
    }

    #[test]
    fn test_resolve_relative_in_root() {
        let dm = DataModel::new(json!({"name": "Alice"}));
        let resolver = PathResolver::new(dm);
        assert_eq!(resolver.resolve_relative("name"), Some(json!("Alice")));
    }

    #[test]
    fn test_resolve_relative_in_collection() {
        let dm = DataModel::new(json!({"items": [{"name": "a"}, {"name": "b"}]}));
        let mut resolver = PathResolver::new(dm);
        resolver.enter_collection("/items", 1);
        assert_eq!(resolver.resolve_relative("name"), Some(json!("b")));
    }

    #[test]
    fn test_resolve_to_absolute() {
        let resolver = PathResolver::new(DataModel::empty());
        assert_eq!(
            resolver.resolve_to_absolute("/items", "name", 0),
            "/items/0/name"
        );
    }

    #[test]
    fn test_resolve_dynamic_literal() {
        let resolver = PathResolver::new(DataModel::empty());
        let dv: DynamicValue<String> = DynamicValue::Literal("hello".into());
        assert_eq!(resolver.resolve_dynamic(&dv).unwrap(), "hello");
    }

    #[test]
    fn test_resolve_dynamic_path() {
        let dm = DataModel::new(json!({"x": 42}));
        let resolver = PathResolver::new(dm);
        let dv: DynamicValue<i64> = DynamicValue::Path { path: "/x".into() };
        assert_eq!(resolver.resolve_dynamic(&dv).unwrap(), 42);
    }
}

File: clients/web-react/src/core/path-resolver.ts
/**
 * 路径解析引擎：根作用域 + 集合作用域（ChildList template）。
 *
 * 对齐 Rust `a2ui-renderer` 的 `PathResolver`：
 * - 绝对路径 `"/…"` 直接解析。
 * - `"@index"` 返回当前集合作用域的索引。
 * - 相对路径在集合作用域下解析为 `"/base/index/relative"`；根作用域下为 `"/relative"`。
 *
 * 解析只读，不修改 Data Model。写回（双向绑定）由上层调用
 * {@link "@/core/data-model".DataModel.applyPointer} 完成。
 */
import type { DataModel } from "@/core/data-model";
import type { Json } from "@/core/json-pointer";

/** 作用域帧：根或集合（数组迭代中的当前项）。 */
export type Scope =
  | { kind: "root" }
  | { kind: "collection"; basePath: string; index: number };

export class PathResolver {
  private readonly dataModel: DataModel;
  private scopeStack: Scope[] = [{ kind: "root" }];

  constructor(dataModel: DataModel) {
    this.dataModel = dataModel;
  }

  /** 解析路径（绝对 / `@index` / 相对），未命中返回 `undefined`。 */
  resolve(path: string): Json | undefined {
    if (path === "@index") {
      const idx = this.currentIndex();
      return idx === undefined ? undefined : idx;
    }
    return this.dataModel.resolvePointer(this.makeAbsolute(path));
  }

  /**
   * 将任意路径规范化为绝对 JSON Pointer（用于依赖登记）。
   * - `"/abs"` 原样返回。
   * - 相对路径：集合作用域 → `"/base/index/rel"`；根作用域 → `"/rel"`。
   * - `"@index"` 无对应数据路径，原样返回。
   */
  makeAbsolute(path: string): string {
    if (path.startsWith("/") || path === "@index") return path;
    const scope = this.currentScope();
    if (scope.kind === "collection") {
      return `${scope.basePath}/${scope.index}/${path}`;
    }
    return `/${path}`;
  }

  /** 当前集合作用域索引；根作用域返回 `undefined`。 */
  currentIndex(): number | undefined {
    const scope = this.currentScope();
    return scope.kind === "collection" ? scope.index : undefined;
  }

  /**
   * 当前作用域的集合帧快照（根帧除外），按从外到内的顺序。
   * 供 SurfaceStore 从解析器状态构造子 NodeRef 的 `Scope`。
   */
  frames(): { basePath: string; index: number }[] {
    const out: { basePath: string; index: number }[] = [];
    for (const s of this.scopeStack) {
      if (s.kind === "collection") out.push({ basePath: s.basePath, index: s.index });
    }
    return out;
  }

  /** 进入集合作用域（数组路径 + 当前项索引）。 */
  enterCollection(basePath: string, index: number): void {
    this.scopeStack.push({ kind: "collection", basePath, index });
  }

  /** 退出集合作用域（根帧不可弹出）。 */
  exitCollection(): void {
    if (this.scopeStack.length > 1) this.scopeStack.pop();
  }

  /** 在给定作用域内执行回调，结束后自动恢复（异常安全）。 */
  withCollection<T>(basePath: string, index: number, fn: () => T): T {
    this.enterCollection(basePath, index);
    try {
      return fn();
    } finally {
      this.exitCollection();
    }
  }

  private currentScope(): Scope {
    return this.scopeStack[this.scopeStack.length - 1];
  }
}
